### PR TITLE
Update eth_estimateGas and eth_call RPC APIs for EIP1559 transaction

### PR DESCRIPTION
**Pull request by matkt**
Signed-off-by: Karim TAAM <karim.t2am@gmail.com>

## PR description

Make `eth_estimateGas` compatible with EIP-1559 transactions. This change also impacts `eth_call` so I made it work for `eth_call` too

CallParams has been modified (https://besu.hyperledger.org/en/stable/Reference/API-Objects/#transaction-call-object) :

There are two new optional fields for EIP1559 transactions. If one of the two is added the second must be too.

- feeCap (optional & Wei)
- gasPremium (optional & Wei)

NB : For non-EIP1559 transactions the calls  (eth_call, eth_estimateGas) work as before

--------

- Request `eth_call` (eip1559 transaction)
```json
{"jsonrpc":"2.0","method":"eth_call","params":[{"feeCap":"0x10", "gasPremium":"0x10", "from":"0xFE3B557E8Fb62b89F4916B721be55cEb828dBd73" , "to":"0x89905dCF6c805Bfb69e91bAbc4596127c4F76Db3", "data":"0xcecdc6aa"}, "latest"],"id":1} 
```
- Response 
```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": "0x9b82d2f38fbdf13006bfa741767f793d917e063392737837b580c1c2b1e0bab3"
}
```
------------
- Request `eth_estimateGas` (eip1559 transaction)
```json 
{"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"feeCap":"0x100000", "gasPremium":"0x10000", "from":"fe3b557e8fb62b89f4916b721be55ceb828dbd73","to":"0xdd37f65db31c107f773e82a4f85c693058fef7a9","value":"0x1"}, "latest"],"id":1}
```
- Response
```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": "0x5208"
}
```

### Test performed

- Goerli - Eth_EstimateGas (eip1559 and  legacy transaction)
- EIP1559 Testnet after the fork - Eth_EstimateGas (eip1559 and  legacy transaction)
- EIP1559 Testnet before the fork - Eth_EstimateGas (eip1559 and  legacy transaction)

- Goerli - Eth_Call (eip1559 and  legacy transaction)
- EIP1559 Testnet after the fork - Eth_Call (eip1559 and  legacy transaction)
- EIP1559 Testnet before the fork - Eth_Call (eip1559 and  legacy transaction)

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

Not updated because EIP1559 is an experimental feature

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java [hunk 1]
```diff
@@ -91,6 +91,8 @@ private JsonCallParameter overrideGasLimitAndPrice(
         callParams.getTo() != null ? callParams.getTo().toString() : null,
         Quantity.create(gasLimit),
         Quantity.create(0L),
+        callParams.getGasPremium().map(Quantity::create).orElse(null),
+        callParams.getFeeCap().map(Quantity::create).orElse(null),
```
Comment (RatanRSur) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java:95]: I think while we're here we might as well fix this issue of `CallParams` and `JsonCallParams`. Take a look at `FilterParameter`, that has non-string types but also has the properties to make it json serializable. If you want to make it a separate PR that works!

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGasTest.java [hunk 2]
```diff
@@ -94,8 +94,22 @@ public void shouldReturnErrorWhenTransientTransactionProcessorReturnsEmpty() {
...
-  public void shouldReturnGasEstimateWhenTransientTransactionProcessorReturnsResultSuccess() {
-    final JsonRpcRequestContext request = ethEstimateGasRequest(callParameter());
+  public void shouldReturnErrorWhenTransientEip1559TransactionProcessorReturnsEmpty() {
+    final JsonRpcRequestContext request = ethEstimateGasRequest(eip1559TransactionCallParameter());
+    when(transactionSimulator.process(
+            eq(modifiedEip1559TransactionCallParameter()), any(OperationTracer.class), eq(1L)))
+        .thenReturn(Optional.empty());
+
+    final JsonRpcResponse expectedResponse =
+        new JsonRpcErrorResponse(null, JsonRpcError.INTERNAL_ERROR);
```
Comment (RatanRSur) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGasTest.java:104]: If `process` returns `empty()`, it could be that the world state is unavailable. As I read it, that means that we're sometimes giving the user `INTERNAL_ERROR` for world state unavailability and sometimes `WORLD_STATE_UNAVAILABLE`. Is that right?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java [hunk 3]
```diff
@@ -102,6 +102,12 @@ public MainnetTransactionValidator(
               String.format("gasPrice is less than the current BaseFee"));
         }
       }
+    } else if (transaction.isEIP1559Transaction()) {
+      return ValidationResult.invalid(
+          TransactionInvalidReason.INVALID_TRANSACTION_FORMAT,
+          String.format(
+              "transaction format is invalid, accepted transaction types are %s",
+              acceptedTransactionTypes.toString()));
```
Comment (RatanRSur) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java:110]: Is this tested anywhere?

**Review by RatanRSur — CHANGES_REQUESTED** [review]
Just want some clarity about how we're doing error messages before I can approve.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGasTest.java (hunk 2 above)
Comment (matkt) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGasTest.java:104]: no because the json rpc method will fail fast thanks to this check without going into the simulator.
https://github.com/hyperledger/besu/blob/83e03706b2f424afd080f18addec228d4f9aafee/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java#L65

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java (hunk 1 above)
Comment (matkt) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java:95]: I will do this in another PR because there are already two things in this PR with `eth_call` and `eth_estimateGas`

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java (hunk 3 above)
Comment (matkt) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java:110]: good catch . I just added a test

**Review by RatanRSur — APPROVED** [review]
